File: 프로그래머스/lv1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.cpp
#include <string>
#include <vector>
using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    
    vector<int> restore_doll;
    
    for(int m = 0; m < moves.size(); m++)
    {
        int lin = moves[m] - 1;
        
        bool isCatch = false;
        
        for(int b = board.size()-1; b >= -1; b--)
        {
            if(b == -1)
            {
                isCatch = true;
                restore_doll.push_back(board[b+1][lin]);
                board[b+1][lin] = 0;
                break;
            }
            
            if(board[b][lin] != 0) continue;
            else
            {
                if(b == board.size()-1) break;
                
                isCatch = true;
                restore_doll.push_back(board[b+1][lin]);
                board[b+1][lin] = 0;
                break;
            }
        }
        
        if(isCatch && restore_doll.size() >= 2)
        {
            if(restore_doll[restore_doll.size()-1] == restore_doll[restore_doll.size()-2])
            {
                restore_doll.pop_back(); restore_doll.pop_back();
                answer += 2;
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        Stack<Integer> stak = new Stack<>();
        int answer = 0;
        
        for(int i = 0; i < moves.length; i++) {
            int x = moves[i]-1;
            
            for(int y = 0; y < board.length; y++) {
                if(board[y][x] == 0) continue;
                
                if(stak.isEmpty()) {
                    stak.push(board[y][x]);
                } else {
                    if(stak.peek() == board[y][x]) {
                        stak.pop();
                        answer += 2;
                    } else {
                        stak.add(board[y][x]);
                    }
                }
                
                board[y][x] = 0;
                break;
            }
            
        }
        
        
        
        return answer;
    }
}